homeassistant/components/generic_water_heater/water_heater.py

```diff
@@ -0,0 +1,228 @@
+            self._current_operation = old_state.state
+
+        temp_sensor = self.hass.states.get(self.sensor_entity_id)
+        if temp_sensor:
+            self._current_temperature = float(temp_sensor.state)
+
+        heater_switch = self.hass.states.get(self.heater_entity_id)
+        if heater_switch and heater_switch.state not in (
+            STATE_UNAVAILABLE,
+            STATE_UNKNOWN,
+        ):
+            self._available = True
+        self.async_write_ha_state()
+
+    async def _async_sensor_changed(self, event):
+        """Handle temperature changes."""
+        new_state = event.data.get("new_state")
+        if new_state is None or new_state.state in (STATE_UNAVAILABLE, STATE_UNKNOWN):
+            # Failsafe
+            _LOGGER.warning(
+                "No Temperature information, entering Failsafe, turning off heater %s",
+                self.heater_entity_id,
+            )
+            await self._async_heater_turn_off()
+            self._current_temperature = None
+        else:
+            self._current_temperature = float(new_state.state)
+
+        await self._async_control_heating()
+
+    @callback
+    def _async_switch_changed(self, event):
+        """Handle heater switch state changes."""
+        new_state = event.data.get("new_state")
+        if new_state is None or new_state.state in (STATE_UNAVAILABLE, STATE_UNKNOWN):
+            self._available = False
+        else:
+            self._available = True
+            if new_state.state == STATE_ON and self._current_operation == STATE_OFF:
+                self._current_operation = new_state.state
+
+        self.async_write_ha_state()
+
+    async def _async_control_heating(self):
+        """Check if we need to turn heating on or off."""
+        if self._current_operation == STATE_OFF or self._current_temperature is None:
+            pass
+        elif (
+            self._current_temperature
+            < self._target_temperature - self._temperature_delta
+        ):
+            _LOGGER.debug("Turning on heater %s", self.heater_entity_id)
+            await self._async_heater_turn_on()
+        else:
+            _LOGGER.debug("Turning off heater %s", self.heater_entity_id)
+            await self._async_heater_turn_off()
+        self.async_write_ha_state()
+
+    async def _async_heater_turn_on(self):
+        """Turn heater toggleable device on."""
+        heater = self.hass.states.get(self.heater_entity_id)
+        if heater is None or heater.state == STATE_ON:
+            return
+
+        data = {ATTR_ENTITY_ID: self.heater_entity_id}
+        await self.hass.services.async_call(
+            HA_DOMAIN, SERVICE_TURN_ON, data, context=self._context
+        )
+
+    async def _async_heater_turn_off(self):
+        """Turn heater toggleable device off."""
+        heater = self.hass.states.get(self.heater_entity_id)
+        if heater is None or heater.state == STATE_OFF:
+            return
```

Comment:
```
# pytest --cov=homeassistant/components/generic_water_heater/ --cov-report term-missing -- tests/components/generic_water_heater/test_*.py
Test session starts (platform: linux, Python 3.8.1, pytest 6.2.2, pytest-sugar 0.9.4)
rootdir: /root/home-assistant, configfile: pyproject.toml
plugins: aiohttp-0.3.0, profiling-1.7.0, parallel-0.1.0, anyio-1.3.0, forked-1.1.3, mock-3.1.1, socket-0.3.5, aresponses-2.0.0, test-groups-1.0.3, requests-mock-1.8.0, reportlog-0.1.2, rerunfailures-9.1.1, cov-2.10.1, sugar-0.9.4, timeout-1.4.2, xdist-2.1.0
collecting ... 
 tests/components/generic_water_heater/test_water_heater.py ✓✓✓✓✓✓                                                                                                  100% ██████████

----------- coverage: platform linux, python 3.8.1-final-0 -----------
Name                                                            Stmts   Miss  Cover   Missing
---------------------------------------------------------------------------------------------
homeassistant/components/generic_water_heater/__init__.py          19      0   100%
homeassistant/components/generic_water_heater/water_heater.py     120      1    99%   223
---------------------------------------------------------------------------------------------
TOTAL                                                             139      1    99%

```

Looks like this line is missing coverage.

Reply:
100% coverage 😎  thank you!